services/spar/test/Test/Spar/Sem/IdPSpec.hs

```diff
@@ -9,64 +10,237 @@ import Imports
+    prop "storeConfig/storeConfig (different keys)" $ prop_storeStore lower
+    prop "storeConfig/storeConfig (same keys)" $ prop_storeStore lower
 
 spec :: Spec
 spec = modifyMaxSuccess (const 1000) $ do
-  propsForInterpreter testInterpreter
+  propsForInterpreter "idPToMem" $ pure . run . idPToMem
+
+getReplacedBy :: Member E.IdP r => SAML.IdPId -> Sem r (Maybe (Maybe SAML.IdPId))
+getReplacedBy idpid = fmap (view $ SAML.idpExtraInfo . IP.wiReplacedBy) <$> E.getConfig idpid
+
+-- | All the constraints we need to generalize properties in this module.
+-- A regular type synonym doesn't work due to dreaded impredicative
+-- polymorphism.
+class
+  (Member E.IdP r, forall z. Show z => Show (f z), forall z. Eq z => Eq (f z)) =>
+  PropConstraints r f
+
+instance
+  (Member E.IdP r, forall z. Show z => Show (f z), forall z. Eq z => Eq (f z)) =>
+  PropConstraints r f
+
+prop_storeStore ::
+  PropConstraints r f =>
+  (forall a. Sem r a -> IO (f a)) ->
+  Property
+prop_storeStore =
+  prepropLaw @'[E.IdP] $ do
+    s <- arbitrary
+    s' <- arbitrary
+    pure
+      ( do
+          E.storeConfig $ s & SAML.idpId .~ s' ^. SAML.idpId
+          E.storeConfig s',
+        do
+          E.storeConfig s'
+      )
+
+prop_storeStoreInterleave ::
+  PropConstraints r f =>
+  (forall a. Sem r a -> IO (f a)) ->
+  Property
+prop_storeStoreInterleave =
+  prepropLaw @'[E.IdP] $ do
+    s <- arbitrary
+    s' <- arbitrary
+    !_ <-
+      if s ^. SAML.idpId == s' ^. SAML.idpId
+        then discard
+        else pure ()
+    pure
+      ( do
+          E.storeConfig s
+          E.storeConfig s',
+        do
+          E.storeConfig s'
+          E.storeConfig s
+      )
+
+prop_storeGet ::
+  PropConstraints r f =>
+  (forall a. Sem r a -> IO (f a)) ->
+  Property
+prop_storeGet =
+  prepropLaw @'[E.IdP] $
+    do
+      s <- arbitrary
+      pure
+        ( do
+            E.storeConfig s
+            E.getConfig $ s ^. idpId,
+          do
+            E.storeConfig s
+            pure (Just s)
+        )
+
+prop_deleteGet ::
+  PropConstraints r f =>
+  (forall a. Sem r a -> IO (f a)) ->
+  Property
+prop_deleteGet =
+  prepropLaw @'[E.IdP] $ do
+    s <- arbitrary
+    pure
+      ( do
+          E.deleteConfig s
+          E.getConfig $ s ^. SAML.idpId,
+        do
+          E.deleteConfig s
+          pure Nothing
+      )
+
+prop_deleteDelete ::
+  PropConstraints r f =>
+  (forall a. Sem r a -> IO (f a)) ->
+  Property
+prop_deleteDelete =
+  prepropLaw @'[E.IdP] $ do
+    s <- arbitrary
+    pure
+      ( do
+          E.deleteConfig s
+          E.deleteConfig s,
+        do
+          E.deleteConfig s
+      )
+
+prop_storeGetByIssuer ::
+  PropConstraints r f =>
+  (forall a. Sem r a -> IO (f a)) ->
+  Property
+prop_storeGetByIssuer =
+  prepropLaw @'[E.IdP] $
+    do
+      s <- arbitrary
+      pure
+        ( do
+            E.storeConfig s
+            E.getIdByIssuerWithoutTeam $ s ^. idpMetadata . edIssuer,
```

Comment:
What about "with team"?  (Extract the team to pass to the getter from s.  If there are two idps with the same issuer and different team, it returns the right one.)

Reply:
If you don't want to implement it, please leave this as `FUTUREWORK`.

Reply:
Maybe we can pair on this one. The semantics weren't immediately clear to me.